immutable/imx-core-sdk
# [DX-907] SDK docs gen

\*\***NOTE: These changes will not be merged**\*\*

# Summary
This PR is to demonstrate the config and output of SDK doc generation using two different tools:
* [TypeDoc](https://typedoc.org/) ([commit](https://github.com/immutable/imx-core-sdk/commit/859e68050fb2f5f89214a2644ef9d6f6adeefa96))
* [ts-docs](https://tsdocs.xyz/) [(commit](https://github.com/immutable/imx-core-sdk/commit/e8e65c6cd0d981066c10edc7a0d438a69e70be3e))

### TypeDoc:
* Config file: `./typedoc.json`
* Output:
    * HTML (`./sdk-docs-html`)
    * Markdown (`./sdk-docs-md`)
    * JSON (`./sdk-docs-json`)

### ts-docs:
* Config file: `tsDocOptions` in `./tsconfig.json`
* Output:
    * HTML (`./tsdocs-html`)
    * JSON (`./tsdocs.json`)

# How to test

Check out this branch and `npm install`. We will be using TypeDoc's and ts-docs' command line tools to generate the files.

## TypeDoc
* The config for TypeDoc [can be stored](https://typedoc.org/guides/options/#options-1) in a `typedoc.json` configuration file
* If compilerOptions is provided in the `typedoc.json`, it will [override the config](https://typedoc.org/guides/options/#compileroptions) in the `tsconfig.json` file
* `"out"` specifies the output folder for HTML or Markdown outputs
* `"json"` specifies the output file for the JSON output (if no JSON output is required, simply remove `"json"` as a key)
* As we have installed and are utilising [typedoc-plugin-markdown](https://github.com/tgreyuk/typedoc-plugin-markdown/tree/master/packages/typedoc-plugin-markdown#readme) to generate Markdown files, which changes the default output of TypeDoc from HTML to Markdown, we need to make sure the  [plugin array](https://github.com/immutable/imx-core-sdk/pull/80/files#diff-6169a21f84f4d19da38a12219c19778338487973f106b9dad94370e66e1c8c1aR8-R10) contains `"none"` if we want the output to be HTML.

### typedoc.json:

#### Generate HTML
```json
{
    "$schema": "https://typedoc.org/schema.json",
    "entryPoints": [
        "./src/index.ts"
    ],
    "out": "sdk-docs-html",
    "plugin": [
        "none"
    ],
    "compilerOptions": {
        "skipLibCheck": true
    }
}
```

#### Generate Markdown
```json
{
    "$schema": "https://typedoc.org/schema.json",
    "entryPoints": [
        "./src/index.ts"
    ],
    "out": "sdk-docs-md",
    "plugin": [
        "typedoc-plugin-markdown"
    ],
    "compilerOptions": {
        "skipLibCheck": true
    }
}
```

#### Generate JSON
```json
{
    "$schema": "https://typedoc.org/schema.json",
    "entryPoints": [
        "./src/index.ts"
    ],
    "json": "sdk-docs-json.json",
    "compilerOptions": {
        "skipLibCheck": true
    }
}
```

### Command to generate documentation based on typedoc.json config:
```shell
npx typedoc --options typedoc.json
```

## ts-docs
* The config for ts-docs [can be stored](https://tsdocs.xyz/pages/Guides/Options.html) as a `"tsdocsOptions"` in the `tsconfig.json` file
* `"out"` specifies the output folder for the HTML output
* `"json"` specifies the output file for the JSON output (if no JSON output is required, simply remove `"json"` as a key)

If you are setting the configuration in the `tsconfig.json` file, it should look like this:
```json
{
  "tsdocsOptions": {
    "entryPoints": [
      "./src/index.ts"
    ],
    "out": "./tsdocs-html",
    "json": "./tsdocs.json"
  }
}
```
### Command to generate documentation using ts-docs:
```shell
npx ts-docs
```

# Why the changes

[Jira](https://immutable.atlassian.net/browse/DX-907) - Investigation into the different tools available for automatically generating SDK docs from this repository.

# Things worth calling out
